firebase-firestore/src/main/java/com/google/firebase/firestore/pipeline/stage.kt

```diff
@@ -0,0 +1,671 @@
+   * @return [FindNearestStage] with specified [distanceField].
+   */
+  fun withDistanceField(distanceField: Field): FindNearestStage =
+    with("distance_field", distanceField)
+
+  /**
+   * Add a field containing the distance to the result.
+   *
+   * @param distanceField The name of the field that will be added to the result.
+   * @return [FindNearestStage] with specified [distanceField].
+   */
+  fun withDistanceField(distanceField: String): FindNearestStage =
+    withDistanceField(field(distanceField))
+}
+
+internal class LimitStage
+internal constructor(private val limit: Int, options: InternalOptions = InternalOptions.EMPTY) :
+  Stage<LimitStage>("limit", options) {
+  override fun self(options: InternalOptions) = LimitStage(limit, options)
+  override fun args(userDataReader: UserDataReader): Sequence<Value> =
+    sequenceOf(encodeValue(limit))
+}
+
+internal class OffsetStage
+internal constructor(private val offset: Int, options: InternalOptions = InternalOptions.EMPTY) :
+  Stage<OffsetStage>("offset", options) {
+  override fun self(options: InternalOptions) = OffsetStage(offset, options)
+  override fun args(userDataReader: UserDataReader): Sequence<Value> =
+    sequenceOf(encodeValue(offset))
+}
+
+internal class SelectStage
+internal constructor(
+  private val fields: Array<out Selectable>,
+  options: InternalOptions = InternalOptions.EMPTY
+) : Stage<SelectStage>("select", options) {
+  override fun self(options: InternalOptions) = SelectStage(fields, options)
+  override fun args(userDataReader: UserDataReader): Sequence<Value> =
+    sequenceOf(encodeValue(fields.associate { it.getAlias() to it.toProto(userDataReader) }))
+}
+
+internal class SortStage
+internal constructor(
+  private val orders: Array<out Ordering>,
+  options: InternalOptions = InternalOptions.EMPTY
+) : Stage<SortStage>("sort", options) {
+  override fun self(options: InternalOptions) = SortStage(orders, options)
+  override fun args(userDataReader: UserDataReader): Sequence<Value> =
+    orders.asSequence().map { it.toProto(userDataReader) }
+}
+
+internal class DistinctStage
+internal constructor(
+  private val groups: Array<out Selectable>,
+  options: InternalOptions = InternalOptions.EMPTY
+) : Stage<DistinctStage>("distinct", options) {
+  override fun self(options: InternalOptions) = DistinctStage(groups, options)
+  override fun args(userDataReader: UserDataReader): Sequence<Value> =
+    sequenceOf(encodeValue(groups.associate { it.getAlias() to it.toProto(userDataReader) }))
+}
+
+internal class RemoveFieldsStage
+internal constructor(
+  private val fields: Array<out Field>,
+  options: InternalOptions = InternalOptions.EMPTY
+) : Stage<RemoveFieldsStage>("remove_fields", options) {
+  override fun self(options: InternalOptions) = RemoveFieldsStage(fields, options)
+  override fun args(userDataReader: UserDataReader): Sequence<Value> =
+    fields.asSequence().map(Field::toProto)
+}
+
+internal class ReplaceStage
```

Comment:
Do you have intent to make all of these internal stages public to support options overrides?

Reply:
Yes.

Does it need to be done before release? Maybe (your call)

I only exposed the stage classes when the API required it, not just future proofing raw options. Exposing the stages will require also adding constructors / create methods that offer same semantics and documentation as Pipeline API. I wasn't going to do this work until the dust settled on options, and it is not critical for release in my opinion. There is always the rawStage.